### PR TITLE
Fix rollback logic to be consistent with old JPA behavior.

### PR DESCRIPTION
**Pull request by Naros**

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java [hunk 1]
```diff
@@ -2363,7 +2363,14 @@ public void afterTransactionCompletion(boolean successful, boolean delayed) {
 		log.tracef( "SessionImpl#afterTransactionCompletion(successful=%s, delayed=%s)", successful, delayed );
 
 		if ( !isClosed() ) {
-			if ( autoClear ) {
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java]: I think both cases should simply do `if ( autoClear || !successful ) { internalCLear(): }` ?

Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java]: Aka, make this entire block:

````
if ( !isClosed() ) {
    if ( autoClear || !successful ) {
        internalClear();
    }
}
````

Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java]: FWIW, I made this change locally and it caused no regression in hibernate-core tests

**Comment by Naros** [comment]
Fixed PR so it does the same for both cases.

**Comment by sebersole** [comment]
I just pushed the same.

**Pull request closed without merging**